File: backend/src/common/utilities/geo.py
from math import radians, sin, cos, sqrt, atan2
from src.common.constants.geo import EARTH_RADIUS_METERS


def distance_meters(lon1: float, lat1: float, lon2: float, lat2: float) -> float:
    phi1, phi2 = radians(lat1), radians(lat2)
    dphi = radians(lat2 - lat1)
    dlambda = radians(lon2 - lon1)
    a = sin(dphi / 2) ** 2 + cos(phi1) * cos(phi2) * sin(dlambda / 2) ** 2
    return EARTH_RADIUS_METERS * 2 * atan2(sqrt(a), sqrt(1 - a))

File: frontend/lib/common/utilities/geo.dart
import 'dart:math';

import 'package:streetballer/common/models/geolocation.dart';

const double _earthRadiusMeters = 6371000.0;

/// Returns the haversine distance in metres between two coordinates.
double haversineMeters(Geolocation a, Geolocation b) {
  final lat1 = a.latitude * pi / 180;
  final lat2 = b.latitude * pi / 180;
  final dLat = (b.latitude - a.latitude) * pi / 180;
  final dLon = (b.longitude - a.longitude) * pi / 180;
  final sinDLat = sin(dLat / 2);
  final sinDLon = sin(dLon / 2);
  final x = sinDLat * sinDLat + cos(lat1) * cos(lat2) * sinDLon * sinDLon;
  return 2 * _earthRadiusMeters * atan2(sqrt(x), sqrt(1 - x));
}
